daffodil-japi/src/main/scala/org/apache/daffodil/japi/infoset/Infoset.scala

```diff
@@ -67,13 +70,29 @@ abstract class InfosetInputter extends SInfosetInputter {
    */
   def getNamespaceURI(): String
 
+  /**
+   * Wrapper function to convert the runtimeProperties Scala Map to a Java Map
+   * and call the getSimpleText function. Proably should not be overridden by
+   * those using the Java API. Instead override getSimpleText(primType, runtimeProperties)
+   * or getSimpleText(primType)
+   */
+  override def getSimpleText(primType: NodeInfo.Kind, runtimeProperties: Map[String,String]): String =
+    getSimpleText(primType, runtimeProperties.asJava)
```

Comment:
I was wondering if it's worth getting rid of this `asJava`? Looking at the implementation, asJava allocates a new `MapWrapper`, which implements the `java.util.AbstractMap` interface to access the underlying Scala `Map`. So by using `asJava` we at least aren't creating a completely new `Map` and whatever allocations are involved with that, but we are still allocating this wrapper class. And we do this even when this Maps are empty (which is going to be the most common case). And even worse, most implementations of this Infoset are not even going to implement this two arg version of getSimpleText, so they'll this allocation will just be unused.

So, I'm wondering if when we create the `val runtimeProperties` in ElementBase.scala, we should just create a Java Map, which scala API users can use without a problem, and we never need to allocate anything extra? We could use the `asJava` method and create the wrapper mentioned above, or we could directly create a JavaMap to avoid that extra allocation and indirection? Any thoughts/concerns with this?

Reply:
`asJava`'s new `MapWrapper` allocation occurs only if a Java API user calls this specific two-arg `getSimpleText` method, correct?  There's a third choice worth adding to the previous two choices then:

1. Create Java Map instead of Scala Map in ElementBase.scala.
2. Continue using Scala Map, calling `asJava` and creating wrapper in this method.
3. Check for empty Scala Map in this method and if so, call Java Collections.emptyMap() instead of calling `asJava` and creating wrapper.  This choice is more efficient because it doesn't create a new wrapper object; it just re-uses an existing empty and immutable object.

I think the third choice might be the best compromise.

Reply:
OK, a Java Map in ElementBase sounds good to me then.